### ISSUE TITLE
InvalidArgumentError Invalid JPEG data, size 4096 for retain.py(image_retrain), Not JPEG file 0x00 0x05

### ISSUE THREAD

**Issue by clarklight** [tensorflow/tensorflow#7434]
Problem: when i run the retrain script with my sub_folder(my photos) shown in the Image retrain tutorial,
https://www.tensorflow.org/how_tos/image_retraining/
 It gives that error, InvalidArgumentError Invalid JPEG data. I took a look at this thread https://github.com/tensorflow/tensorflow/issues/4009

And changed the tf.image.decode_jpeg to tf.image.decode_image, but no luck(i also made sure all the photos are jpeg). I am wondering the retrain script for the tutorial is there a Max-size set for the images? Because i couldnt find if there is a max-size, or if its still some problem with format of the photos i put in. 

I have ran the flower_photos example with no problem, running on Ubuntu 16.04, CUDA 8.0. 

Edit: I am using the flower_photos dataset from the tensorflow website, i used a USB stick to transfer the file to the window machine running virtual box Ubuntu 16.04, i deleted 4 sub_folders(only training daisy and dandelion), then the bottleneck process went through about 912 photos and suddenly a daisy photo had the same Invalid JPEG error. I am wondering if its because of the photos transfered through a USB from a Mac to Window then onto a virtual box causes JPEG to corrupt?

**Comment by clarklight** [comment]
I have managed to locate the root of my problem, the transfer of JPEGS dataset via USB will corrupt the file, so that when you decode it, the decoder will think its not JPEG calling a Invalid JPEG 0x00 0x05.

**Comment by ktdiedrich** [comment]
I had a similar error even though my images appear to be valid JPEG images. Like the previous error my Tensorflow error was:
InvalidArgumentError (see above for traceback): Invalid JPEG data, size 4096
         [[Node: DecodeJpeg = DecodeJpeg[acceptable_fraction=1, channels=3, dct_method="", fancy_upscaling=true, ratio=1, try_recover_truncated=false, _device="/job:localhost/replica:0/task:0/cpu:0"](_recv_DecodeJpeg/contents_0)]]
I checked the JPEG files with 
file *.jpg
They are all: JPEG image data, JFIF standard 1.01
With vi -b 01-007.jpg   The files start: 
<ff><d8><ff><e0>^@^PJFIF^@^A^A^A^A,
Which seems to match "Valid JFIF files begin with FF D8 FF E0 ?? ?? 'J' 'F' 'I' 'F' 00." from Wikipedia: http://fileformats.archiveteam.org/wiki/JFIF 
All the images decode without error using a Tensorflow call
with tf.Graph().as_default():
        image_contents = tf.read_file(filePath)
        image = tf.image.decode_jpeg(image_contents, channels=channels)
        init_op = tf.tables_initializer()
        with tf.Session() as sess:
            sess.run(init_op)
            decoded = sess.run(image)
So I don't know what causing the error.

**Comment by clovking** [comment]
Does the problem have been solved?

**Comment by clovking** [comment]
@clarklight ,  I meet a silimar question with you when i train models on google TensorFlow Object Detection API, May i ask if the question has been solved? what's the problem and how to solved? thank you!

**Comment by clarklight** [comment]
Hi,not sure if this manage to response back to the thread. That problem
occurs when the Jpeg starting pixel breaks.....like the starting bits is
missing. That occurs for me when I transferred images from a system to
another via USB. You have to check if the base64 is still 100% intact, the
image will view ok, but you will see the base64 is different. If the image
is correct then it will train correctly.

**Comment by clarklight** [comment]
For me anyway the jpeg looked like it was correct, the format looked
correct. I checked everything,I even open it to view the image, everything
looked right, but for some reason for me when the images is moved from a
folder to another some transfer method causes jpeg to loss some bits from
it's base64 string.......no idea why. When I compared the Actual base64
then I saw the string was slightly different.......then I tried a different
transfer method, when the base64 string is 100% the same.....then it
worked. No idea why, I think the Tensorflow has to mount the image into the
image parser in C++ then read it, so basically the system couldn't mount
the image to parse and gave that error, if you guys base64 is correct or
file format is correct, maybe take a look at the log to see if the image is
even getting read correctly.

**Comment by clovking** [comment]
Thank you for your patient and detailed explanation. I will try it as you say. Thank you!!!







At 2018-02-23 19:30:01, "clarklight" <notifications@github.com> wrote:
For me anyway the jpeg looked like it was correct, the format looked
correct. I checked everything,I even open it to view the image, everything
looked right, but for some reason for me when the images is moved from a
folder to another some transfer method causes jpeg to loss some bits from
it's base64 string.......no idea why. When I compared the Actual base64
then I saw the string was slightly different.......then I tried a different
transfer method, when the base64 string is 100% the same.....then it
worked. No idea why, I think the Tensorflow has to mount the image into the
image parser in C++ then read it, so basically the system couldn't mount
the image to parse and gave that error, if you guys base64 is correct or
file format is correct, maybe take a look at the log to see if the image is
even getting read correctly.

**Comment by clarklight** [comment]
Something funny happened, I used to run my Tensorflow on Linux and Unix
system. Today was my first time to install and run on a window
device.....and I ran into a problem with retraining a model as well. But
it's slightly different. It kept complaining about flower_photos not found.
As if the directory isn't there.....but I made sure it is........not sure
what's the problem still.....

**Comment by clovking** [comment]
It , I run google object detection API, shows the error log such as : 

Input file read error
INFO:tensorflow:Error reported to Coordinator: <class 'tensorflow.python.framework.errors_impl.InvalidArgumentError'>, Invalid JPEG data or crop window, data size 166476
[[Node: case/If_0/decode_image/cond_jpeg/DecodeJpeg = DecodeJpeg[acceptable_fraction=1, channels=3, dct_method="", fancy_upscaling=true, ratio=1, try_recover_truncated=false, _device="/job:localhost/replica:0/task:0/device:CPU:0"](case/If_0/ decode_image/cond_jpeg/cond_png/DecodePng/Switch:1, ^case/Assert/AssertGuard/Merge, ^case/If_0/decode_image/cond_jpeg/Assert/ Assert)]]. 

It is found that a few pictures are invalid format when I use tf.image.decode_jpeg() function to detect and the same error upon is found. Just discard this invalid pictures. The function will consume a lot of time.
Of course, maybe the question , you and me, is not the same question, just give the method which is valid for me.
Thank you for your reply.

发送自 Windows 10 版邮件应用

发件人: clarklight
发送时间: 2018年3月14日 2:05
收件人: tensorflow/tensorflow
抄送: clovking; Comment
主题: Re: [tensorflow/tensorflow] InvalidArgumentError Invalid JPEG data,size 4096 for retain.py(image_retrain), Not JPEG file 0x00 0x05 (#7434)

Something funny happened, I used to run my Tensorflow on Linux and Unix
system. Today was my first time to install and run on a window
device.....and I ran into a problem with retraining a model as well. But
it's slightly different. It kept complaining about flower_photos not found.
As if the directory isn't there.....but I made sure it is........not sure
what's the problem still.....

**Comment by milad-4274** [comment]
I faced similar problem. there is a problem in some of your training data. you can use code below to check which jpeg image is corrupted and delete it.

```
from struct import unpack
import os


marker_mapping = {
    0xffd8: "Start of Image",
    0xffe0: "Application Default Header",
    0xffdb: "Quantization Table",
    0xffc0: "Start of Frame",
    0xffc4: "Define Huffman Table",
    0xffda: "Start of Scan",
    0xffd9: "End of Image"
}


class JPEG:
    def __init__(self, image_file):
        with open(image_file, 'rb') as f:
            self.img_data = f.read()
    
    def decode(self):
        data = self.img_data
        while(True):
            marker, = unpack(">H", data[0:2])
            # print(marker_mapping.get(marker))
            if marker == 0xffd8:
                data = data[2:]
            elif marker == 0xffd9:
                return
            elif marker == 0xffda:
                data = data[-2:]
            else:
                lenchunk, = unpack(">H", data[2:4])
                data = data[2+lenchunk:]            
            if len(data)==0:
                break        


bads = []

for img in tqdm(images):
  image = osp.join(root_img,img)
  image = JPEG(image) 
  try:
    image.decode()   
  except:
    bads.append(img)


for name in bads:
  os.remove(osp.join(root_img,name))
```
I used [yasoob](https://yasoob.me/posts/understanding-and-writing-jpeg-decoder-in-python/) script to decode jpeg image.
